invalid quality_gates migration can brick gsd.db; artifact retries also look like stuck loops in forensics

### Bug type

Regression (worked before, now fails)

### Summary

  ```md
    ## Summary

    Two related issues showed up during auto-mode investigation:

    1. **DB migration bug:** existing `.gsd/gsd.db` can fail to open with errors like `no such column: scope`, causing degraded mode and auto bootstrap failure.
    2. **Forensics/journaling bug:** artifact-verification retries are reported as generic stuck loops because the retry path is not journaled distinctly.

    ## Root Cause

    ### 1) Invalid SQLite DDL in `quality_gates` migration

    `gsd-db.js:846-857` creates `quality_gates` with:

    ```js
    PRIMARY KEY (milestone_id, slice_id, gate_id, COALESCE(task_id, ''))
  ```

  SQLite does not allow expressions inside table PRIMARY KEY definitions.

  This can cause migrateSchema() / openDatabase() to fail for DBs migrating through schema v12, which then bubbles into bootstrap failure via:

  - auto-start.js:54-62
  - auto-start.js:587-597

  Once migration is broken or partial, newer code assumes scope exists in tables like quality_gates / assessments, leading to startup errors such as:

  - no such column: scope

  ### 2) Artifact retries are indistinguishable from generic stuck loops

  When a unit finishes but expected artifacts are missing:

  - auto-post-unit.js:705-712 sets pendingVerificationRetry and returns retry
  - auto/loop.js treats finalize "continue" as finishTurn("retry") + immediate redispatch

  This is valid behavior, but no dedicated journal event is emitted for artifact-verification retries, so forensics only sees repeated dispatch of the same unit after:

  - unit-end status:"completed", artifactVerified:false

  That gets surfaced as a stuck loop even when the loop is coming from the bounded verification retry path.

  Expected Behavior

  - quality_gates migration should use valid SQLite schema.
  - startup errors should distinguish migration failure from provider/corruption problems.
  - artifact-verification retries should emit a dedicated journal event so forensics can classify them correctly.

  Suggested Fix

  ### DB

  Replace the invalid PK with a valid non-expression key, e.g.:

  ```sql
    task_id TEXT NOT NULL DEFAULT ''
    PRIMARY KEY (milestone_id, slice_id, gate_id, task_id)
  ```

  Then add a repair migration that rebuilds quality_gates from old rows using COALESCE(task_id, '').

  Also add required-column validation for:
  - quality_gates.scope
  - assessments.scope

  before treating the DB as healthy.

  ### Journaling / Forensics

  Emit a dedicated journal event before verification-triggered redispatch, e.g.:

  - artifact-verification-retry

  with:
  - unitType
  - unitId
  - attempt
  - expected artifact path
  - failure context

  Then exclude these from generic stuck-loop reporting until retry budget is exhausted.

  Environment

  - GSD version: 2.76.0
  - Observed during auto-mode on M007/S01/T01

  Evidence

  ### Bootstrap failure path

  - auto-start.js:62 logs DB open failure
  - auto-start.js:596-597 hard-stops auto-mode if DB file exists but DB is unavailable

  ### Invalid migration

  - gsd-db.js:846-857

  ### Retry-loop evidence

  Journal showed repeated sequences like:

  - unit-end ... status:"completed", artifactVerified:false
  - immediate redispatch of the same unit

  Seen for:
  - research-slice/M007/S01
  - plan-slice/M007/S01
  - execute-task/M007/S01/T01

  ────────────────────────────────────────────────────────────────────────────────

  Auto-generated from local forensic analysis with redacted paths.

### Steps to reproduce

  ```md
    ## Steps to Reproduce

    1. Start from a project with an existing `.gsd/gsd.db` created by an older GSD version that still needs schema migration
  through the `quality_gates` migration path.
    2. Upgrade to a build containing the current `gsd-db.js` migration logic.
    3. Run `/gsd auto` on that project.
    4. During bootstrap, GSD attempts to open and migrate the existing database.
    5. Migration fails in the `quality_gates` schema step because the table definition uses an expression inside the `PRIMARY
  KEY`:
       - `PRIMARY KEY (milestone_id, slice_id, gate_id, COALESCE(task_id, ''))`
    6. Observe startup warnings/errors such as:
       - `gsd-db: failed to open existing database: no such column: scope`
       - `DB unavailable — using filesystem state derivation (degraded mode)`
       - `SQLite database exists but failed to open. Auto-mode cannot proceed...`

    ### For the retry / stuck-loop symptom

    1. Run `/gsd auto` on a unit whose execution completes but does not produce its expected artifact.
    2. Let post-unit verification fail with `artifactVerified:false`.
    3. Observe that auto-mode immediately redispatches the same unit.
    4. Check the journal and note the sequence:
       - `unit-end ... status:"completed", artifactVerified:false`
       - next iteration dispatches the same unit again
    5. Run `/gsd forensics` and observe that these retries are surfaced as repeated dispatch / stuck-loop anomalies rather than a
  distinct artifact-verification retry class.
  ```

### Expected behavior

    GSD should open the exisiting DB and continue

### Actual behavior

gsd stops auto mode

### GSD version

v2.76.0

### Node.js version

v24.13.0 

### Operating system

Windows 11

### Install method

npm global install

### Affected area

Auto-mode / dispatch loop

### Model and provider

openai/gpt-5.4

### Logs, screenshots, and evidence

```shell

```

### Impact

_No response_

### Additional information

_No response_

# GSD Forensic Report

**Generated:** 2026-04-20T07:51:00.004Z
**GSD Version:** 2.76.0
**Active Milestone:** M007
**Active Slice:** S01
**Active Worktree:** none

## Problem Description

invalid quality_gates migration can brick gsd.db; artifact retries also look like stuck loops in forensics

## Anomalies Detected (20)

### [WARNING] stuck-loop: Unit complete-slice/M003/S01 was dispatched 2 times
- Unit: complete-slice/M003/S01
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [WARNING] stuck-loop: Unit research-slice/M004/S04 was dispatched 2 times
- Unit: research-slice/M004/S04
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [WARNING] stuck-loop: Unit plan-slice/M004/S06 was dispatched 2 times
- Unit: plan-slice/M004/S06
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [ERROR] stuck-loop: Unit research-slice/M005/S05 was dispatched 4 times
- Unit: research-slice/M005/S05
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [WARNING] stuck-loop: Unit complete-milestone/M005 was dispatched 2 times
- Unit: complete-milestone/M005
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [WARNING] stuck-loop: Unit execute-task/M006/S05/T02 was dispatched 2 times
- Unit: execute-task/M006/S05/T02
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [ERROR] stuck-loop: Unit execute-task/M006/S05/T03 was dispatched 7 times
- Unit: execute-task/M006/S05/T03
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [ERROR] stuck-loop: Unit execute-task/M006/S05/T04 was dispatched 4 times
- Unit: execute-task/M006/S05/T04
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [WARNING] stuck-loop: Unit complete-milestone/M006 was dispatched 2 times
- Unit: complete-milestone/M006
- Repeated dispatch suggests the unit completed but its artifacts weren't verified, or the state machine kept returning it.

### [ERROR] stuck-loop: Unit research-slice/M007/S01 was dispatched 21 times
- Unit: research-slice/M007/S01
- Repeated dispatch within the same auto session suggests the unit completed but its artifacts were not verified, or the state machine kept returning it. Cross-session recovery runs are ignored.

### [WARNING] stuck-loop: Unit plan-slice/M007/S01 was dispatched 2 times
- Unit: plan-slice/M007/S01
- Repeated dispatch within the same auto session suggests the unit completed but its artifacts were not verified, or the state machine kept returning it. Cross-session recovery runs are ignored.

### [ERROR] stuck-loop: Unit execute-task/M007/S01/T01 was dispatched 3 times
- Unit: execute-task/M007/S01/T01
- Repeated dispatch within the same auto session suggests the unit completed but its artifacts were not verified, or the state machine kept returning it. Cross-session recovery runs are ignored.

### [WARNING] cost-spike: $2.66 vs $0.798 average for research-slice
- Unit: research-slice/M002/S03
- Unit research-slice/M002/S03 cost 3.3x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $2.81 vs $0.798 average for research-slice
- Unit: research-slice/M004/S05
- Unit research-slice/M004/S05 cost 3.5x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $3.65 vs $0.941 average for plan-slice
- Unit: plan-slice/M004/S05
- Unit plan-slice/M004/S05 cost 3.9x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $3.32 vs $0.798 average for research-slice
- Unit: research-slice/M004/S09
- Unit research-slice/M004/S09 cost 4.2x the average. May indicate excessive retries or large context.

### [WARNING] cost-spike: $2.45 vs $0.798 average for research-slice
- Unit: research-slice/M005/S04
- Unit research-slice/M005/S04 cost 3.1x the average. May indicate excessive retries or large context.

### [WARNING] error-trace: 1 error(s) in execute/task-M007-S01-T01
- Unit: execute/task-M007-S01-T01
- 
> zdeploy-server@1.0.0 test
> vitest run wakePolicyResolver

[33mThe CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.[39m

[7m[1m[36m RUN [39m[22m[27m [36mv1.6.1[39m [90mC:/Projects/zdeploy/se

### [WARNING] error-trace: 6 error(s) in execute/task-M007-S01-T01
- Unit: execute/task-M007-S01-T01
- Unknown command: "build"


Did you mean this?
  npm run build # run the "build" package script
To see a list of supported npm commands, run:
  npm help


Command exited with code 1

> zdeploy-server@1.0.0 build
> tsc

src/routes/devices.ts(318,14): error TS2552: Cannot find name 'wakePolicyScope'. Did you mean 'wakePolicy'?
src/routes/devices.ts(319,17): error TS2552: Cannot find name 'wakePolicyScope'. Did you mean 'wakePolicy'?
src/routes/standort.ts(8,10): error TS2440
npm error Missing script: "test"
npm error
npm error To see a list of scripts, run:
npm error   npm run
npm error A complete log of this run can be found in: ~\AppData\Local\npm-cache\_logs\2026-04-20T06_51_18_055Z-debug-0.log


Command exited with code 1

### [WARNING] error-trace: 13 error(s) in execute/task-M007-S01-T01
- Unit: execute/task-M007-S01-T01
- Found 2 occurrences of the text in server/src/routes/standort.ts. The text must be unique. Please provide more context to make it unique.
Found 3 occurrences of the text in server/src/routes/devices.ts. The text must be unique. Please provide more context to make it unique.
Found 2 occurrences of the text in server/src/routes/standort.ts. The text must be unique. Please provide more context to make it unique.

## Recent Units

| Type | ID | Cost | Duration | Model |
|------|-----|------|----------|-------|
| execute-task | M007/S01/T01 | $0.704 | 1m 45s | gpt-5.4 |
| execute-task | M007/S01/T01 | $0.590 | 2m 15s | gpt-5.4 |
| execute-task | M007/S01/T01 | $0.960 | 4m 2s | gpt-5.4 |
| execute-task | M007/S01/T01 | $2.14 | 14m 53s | gpt-5.2 |
| plan-slice | M007/S01 | $0.697 | 3m 34s | gpt-5.2 |
| plan-slice | M007/S01 | $0.522 | 1m 45s | gpt-5.2 |
| research-slice | M007/S01 | $0.511 | 2m 50s | gpt-5.2 |
| research-slice | M007/S01 | $0.742 | 3m 17s | gpt-5.2 |
| research-slice | M007/S01 | $0.0000 | 6s | groq/compound |
| research-slice | M007/S01 | $0.0000 | 6s | groq/compound |

## Activity Log Traces (last 5)

### execute/task-M007-S01-T01 (seq 192)
- Tool calls: 33
- Files written: 0
- Errors: 0
- Last reasoning: Recorded: `wakePolicy` passed cleanly.

Updated T01 verification state:
- ✅ `npm --prefix server test -- siteFilter`
- ✅ `npm --prefix server test -- wakePolicy`
- ✅ `npm --prefix server run build`
- 

### execute/task-M007-S01-T01 (seq 191)
- Tool calls: 32
- Files written: 1
- Errors: 1
- Last reasoning: st files found`; resolver test file does not exist yet)
- `npm --prefix server test -- wakePolicy` → ✅ pass
- `npm --prefix server test -- siteFilter` → ✅ pass
- `npm --prefix server run build` → ✅ pa

### execute/task-M007-S01-T01 (seq 190)
- Tool calls: 61
- Files written: 8
- Errors: 6
- Last reasoning: Task T01 complete.

### execute/task-M007-S01-T01 (seq 189)
- Tool calls: 105
- Files written: 6
- Errors: 13
- Last reasoning: Task T01 complete.

### plan/slice-M007-S01 (seq 188)
- Tool calls: 28
- Files written: 4
- Errors: 0
- Last reasoning: Slice S01 planned.

## Doctor Issues

- [WARN] project | large_planning_file | 1 planning file(s) exceed 100KB — largest: C:\Projects\zdeploy\.gsd\milestones\M001\M001-RESEARCH.md (149KB). Large files cause LLM context pressure. | file: C:\Projects\zdeploy\.gsd\milestones\M001\M001-RESEARCH.md | fixable: no
- [WARN] environment | env_docker | Project has Docker files but docker is not installed | fixable: no
- [WARN] environment | env_git_remote | Git remote 'origin' is unreachable — Remote: http://forgeddy_admin:c12dd89fd1f17c053abec1947b2f95dccdc637ba@192.168.77.200:3000/forgeddy_admin/zdeploy.git | fixable: no
- [WARN] project | db_unavailable | Database unavailable — using filesystem state derivation (degraded mode). State queries may be slower and less reliable. | file: .gsd/gsd.db | fixable: no
- [WARN] project | provider_key_missing | Anthropic (Claude) — not configured — Run /gsd keys to authenticate | fixable: no
- [WARN] M001/S01 | delimiter_in_title | Slice M001/S01 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S01/T02 | task_done_must_haves_not_verified | Task T02 has 3 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S01/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S01/T03 | task_done_must_haves_not_verified | Task T03 has 4 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S01/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S01/T04 | task_done_must_haves_not_verified | Task T04 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S01/tasks/T04-SUMMARY.md | fixable: no
- [WARN] M001/S02 | delimiter_in_title | Slice M001/S02 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S02/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S02/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S02/T02 | task_done_must_haves_not_verified | Task T02 has 4 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S02/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S02/T03 | task_done_must_haves_not_verified | Task T03 has 5 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S02/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S03 | delimiter_in_title | Slice M001/S03 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S03/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S03/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S03/T02 | task_done_must_haves_not_verified | Task T02 has 7 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S03/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S03/T03 | task_done_must_haves_not_verified | Task T03 has 8 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S03/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S04 | delimiter_in_title | Slice M001/S04 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S04/T01 | task_done_must_haves_not_verified | Task T01 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S04/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S04/T02 | task_done_must_haves_not_verified | Task T02 has 5 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S04/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S04/T03 | task_done_must_haves_not_verified | Task T03 has 7 must-haves but summary addresses only 5 | file: .gsd/milestones/M001/slices/S04/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S05 | delimiter_in_title | Slice M001/S05 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S05/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S05/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S05/T02 | task_done_must_haves_not_verified | Task T02 has 5 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S05/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S05/T03 | task_done_must_haves_not_verified | Task T03 has 5 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S05/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S05/T04 | task_done_must_haves_not_verified | Task T04 has 5 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S05/tasks/T04-SUMMARY.md | fixable: no
- [WARN] M001/S06 | delimiter_in_title | Slice M001/S06 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S06/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S06/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S06/T02 | task_done_must_haves_not_verified | Task T02 has 6 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S06/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S06/T04 | task_done_must_haves_not_verified | Task T04 has 8 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S06/tasks/T04-SUMMARY.md | fixable: no
- [WARN] M001/S07 | delimiter_in_title | Slice M001/S07 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S07/T01 | task_done_must_haves_not_verified | Task T01 has 5 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S07/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S07/T02 | task_done_must_haves_not_verified | Task T02 has 8 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S07/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S07/T03 | task_done_must_haves_not_verified | Task T03 has 6 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S07/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S07/T04 | task_done_must_haves_not_verified | Task T04 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S07/tasks/T04-SUMMARY.md | fixable: no
- [WARN] M001/S08 | delimiter_in_title | Slice M001/S08 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S08/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S08/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S08/T02 | task_done_must_haves_not_verified | Task T02 has 5 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S08/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S08/T03 | task_done_must_haves_not_verified | Task T03 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S08/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S09 | delimiter_in_title | Slice M001/S09 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S09/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S09/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S09/T02 | task_done_must_haves_not_verified | Task T02 has 6 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S09/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S09/T03 | task_done_must_haves_not_verified | Task T03 has 5 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S09/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S09/T04 | task_done_must_haves_not_verified | Task T04 has 3 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S09/tasks/T04-SUMMARY.md | fixable: no
- [WARN] M001/S10 | delimiter_in_title | Slice M001/S10 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S10/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S10/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S10/T02 | task_done_must_haves_not_verified | Task T02 has 10 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S10/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S10/T03 | task_done_must_haves_not_verified | Task T03 has 8 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S10/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S11 | delimiter_in_title | Slice M001/S11 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S11/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S11/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S11/T02 | task_done_must_haves_not_verified | Task T02 has 5 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S11/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S11/T03 | task_done_must_haves_not_verified | Task T03 has 8 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S11/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S12 | delimiter_in_title | Slice M001/S12 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S12/T01 | task_done_must_haves_not_verified | Task T01 has 3 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S12/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S13 | delimiter_in_title | Slice M001/S13 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S13/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 5 | file: .gsd/milestones/M001/slices/S13/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S14 | delimiter_in_title | Slice M001/S14 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S15 | delimiter_in_title | Slice M001/S15 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S16 | delimiter_in_title | Slice M001/S16 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S17 | delimiter_in_title | Slice M001/S17 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S18 | delimiter_in_title | Slice M001/S18 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S19 | delimiter_in_title | Slice M001/S19 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S20 | delimiter_in_title | Slice M001/S20 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S21 | delimiter_in_title | Slice M001/S21 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S22 | delimiter_in_title | Slice M001/S22 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S23 | delimiter_in_title | Slice M001/S23 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S24 | delimiter_in_title | Slice M001/S24 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S25 | delimiter_in_title | Slice M001/S25 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S26 | delimiter_in_title | Slice M001/S26 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S27 | delimiter_in_title | Slice M001/S27 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S28 | delimiter_in_title | Slice M001/S28 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S29 | delimiter_in_title | Slice M001/S29 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S30 | delimiter_in_title | Slice M001/S30 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S31 | delimiter_in_title | Slice M001/S31 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S32 | delimiter_in_title | Slice M001/S32 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S33 | delimiter_in_title | Slice M001/S33 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S34 | delimiter_in_title | Slice M001/S34 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S35 | delimiter_in_title | Slice M001/S35 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S36 | delimiter_in_title | Slice M001/S36 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S37 | delimiter_in_title | Slice M001/S37 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S38 | delimiter_in_title | Slice M001/S38 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S39 | delimiter_in_title | Slice M001/S39 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S40 | delimiter_in_title | Slice M001/S40 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S41 | delimiter_in_title | Slice M001/S41 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S42 | delimiter_in_title | Slice M001/S42 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S43 | delimiter_in_title | Slice M001/S43 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S43/T01 | task_done_must_haves_not_verified | Task T01 has 3 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S43/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S43/T02 | task_done_must_haves_not_verified | Task T02 has 4 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S43/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S44 | delimiter_in_title | Slice M001/S44 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S44/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S44/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S44/T02 | task_done_must_haves_not_verified | Task T02 has 6 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S44/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S45 | delimiter_in_title | Slice M001/S45 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S45/T01 | task_done_must_haves_not_verified | Task T01 has 5 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S45/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S45/T02 | task_done_must_haves_not_verified | Task T02 has 4 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S45/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S46 | delimiter_in_title | Slice M001/S46 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S46/T01 | task_done_must_haves_not_verified | Task T01 has 6 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S46/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S46/T02 | task_done_must_haves_not_verified | Task T02 has 8 must-haves but summary addresses only 6 | file: .gsd/milestones/M001/slices/S46/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S46/T03 | task_done_must_haves_not_verified | Task T03 has 4 must-haves but summary addresses only 1 | file: .gsd/milestones/M001/slices/S46/tasks/T03-SUMMARY.md | fixable: no
- [WARN] M001/S47 | delimiter_in_title | Slice M001/S47 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S47/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 3 | file: .gsd/milestones/M001/slices/S47/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S47/T02 | task_done_must_haves_not_verified | Task T02 has 7 must-haves but summary addresses only 4 | file: .gsd/milestones/M001/slices/S47/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M001/S48 | delimiter_in_title | Slice M001/S48 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M001/M001-ROADMAP.md | fixable: no
- [WARN] M001/S48/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 2 | file: .gsd/milestones/M001/slices/S48/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M001/S48/T02 | task_done_must_haves_not_verified | Task T02 has 3 must-haves but summary addresses only 0 | file: .gsd/milestones/M001/slices/S48/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M004/S03/T02 | task_done_must_haves_not_verified | Task T02 has 10 must-haves but summary addresses only 9 | file: .gsd/milestones/M004/slices/S03/tasks/T02-SUMMARY.md | fixable: no
- [WARN] M005/S01 | delimiter_in_title | Slice M005/S01 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M005/M005-ROADMAP.md | fixable: no
- [WARN] M005/S02 | delimiter_in_title | Slice M005/S02 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M005/M005-ROADMAP.md | fixable: no
- [WARN] M005/S03 | delimiter_in_title | Slice M005/S03 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M005/M005-ROADMAP.md | fixable: no
- [WARN] M005/S03/T01 | task_done_must_haves_not_verified | Task T01 has 5 must-haves but summary addresses only 4 | file: .gsd/milestones/M005/slices/S03/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M005/S04 | delimiter_in_title | Slice M005/S04 title contains em/en dash (— or –), which conflict with GSD state document delimiters. Rename the slice to remove these characters to prevent state corruption. | file: .gsd/milestones/M005/M005-ROADMAP.md | fixable: no
- [WARN] M005/S05/T01 | task_done_must_haves_not_verified | Task T01 has 4 must-haves but summary addresses only 3 | file: .gsd/milestones/M005/slices/S05/tasks/T01-SUMMARY.md | fixable: no
- [WARN] M006 | orphaned_slice_directory | Directory "parallel-research" exists in M006/slices/ but is not referenced in the roadmap | file: .gsd/milestones/M006/slices/parallel-research | fixable: no

## Activity Log Metadata

- Files: 192
- Total size: 32557.2 KB
- Oldest: 001-research-slice-M001-S25.jsonl
- Newest: 192-execute-task-M007-S01-T01.jsonl

## Journal Summary

- Total entries: 961
- Distinct flows (iterations): 47
- Daily files: 10
- Date range: 2026-03-27T06:27:44.233Z — 2026-04-20T06:57:28.684Z

### Event Type Distribution

| Event Type | Count |
|------------|-------|
| iteration-start | 46 |
| dispatch-match | 45 |
| unit-start | 44 |
| unit-end | 37 |
| iteration-end | 10 |
| worktree-merge-start | 1 |
| terminal | 1 |

### Recent Journal Events (last 20)

- 2026-04-20T06:34:03.216Z [unit-start] flow=4ff7e86e unit=M007/S01/T01
- 2026-04-20T06:47:43.624Z [unit-end] flow=4ff7e86e unit=M007/S01/T01
- 2026-04-20T06:48:56.262Z [unit-end] flow=4ff7e86e unit=M007/S01/T01
- 2026-04-20T06:49:20.763Z [iteration-start] flow=e59cf353
- 2026-04-20T06:49:20.944Z [dispatch-match] flow=e59cf353 rule=executing → execute-task unit=M007/S01/T01
- 2026-04-20T06:49:21.024Z [unit-start] flow=e59cf353 unit=M007/S01/T01
- 2026-04-20T06:53:23.716Z [unit-end] flow=e59cf353 unit=M007/S01/T01
- 2026-04-20T06:53:25.289Z [iteration-end] flow=e59cf353
- 2026-04-20T06:53:25.293Z [iteration-start] flow=b56c64f9
- 2026-04-20T06:53:25.627Z [dispatch-match] flow=b56c64f9 rule=executing → execute-task unit=M007/S01/T01
- 2026-04-20T06:53:25.706Z [unit-start] flow=b56c64f9 unit=M007/S01/T01
- 2026-04-20T06:55:40.775Z [unit-end] flow=b56c64f9 unit=M007/S01/T01
- 2026-04-20T06:55:42.280Z [iteration-end] flow=b56c64f9
- 2026-04-20T06:55:42.283Z [iteration-start] flow=07c46af4
- 2026-04-20T06:55:42.571Z [dispatch-match] flow=07c46af4 rule=executing → execute-task unit=M007/S01/T01
- 2026-04-20T06:55:42.656Z [unit-start] flow=07c46af4 unit=M007/S01/T01
- 2026-04-20T06:57:27.909Z [unit-end] flow=07c46af4 unit=M007/S01/T01
- 2026-04-20T06:57:28.466Z [iteration-end] flow=07c46af4
- 2026-04-20T06:57:28.468Z [iteration-start] flow=3daf65f4
- 2026-04-20T06:57:28.684Z [dispatch-match] flow=3daf65f4 rule=executing → execute-task unit=M007/S01/T01

--- a/src/resources/extensions/gsd/auto/phases.ts
+++ b/src/resources/extensions/gsd/auto/phases.ts
@@ -1928,6 +1928,20 @@ export async function runFinalize(
       debugLog("autoLoop", { phase: "sidecar-artifact-retry-skipped", iteration: ic.iteration });
     } else {
       // s.pendingVerificationRetry was set by postUnitPreVerification.
+      // Emit a dedicated journal event so forensics can distinguish bounded
+      // verification retries from genuine stuck-loop dispatch repetitions (#4540).
+      const retryInfo = s.pendingVerificationRetry;
+      deps.emitJournalEvent({
+        ts: new Date().toISOString(),
+        flowId: ic.flowId,
+        seq: ic.nextSeq(),
+        eventType: "artifact-verification-retry",
+        data: {
+          unitType: preUnitSnapshot?.type,
+          unitId: retryInfo?.unitId,
+          attempt: retryInfo?.attempt,
+        },
+      });
       // Continue the loop — next iteration will inject the retry context into the prompt.
       debugLog("autoLoop", { phase: "artifact-verification-retry", iteration: ic.iteration });
       return { action: "continue" };

--- a/src/resources/extensions/gsd/gsd-db.ts
+++ b/src/resources/extensions/gsd/gsd-db.ts
@@ -180,7 +180,7 @@ function openRawDb(path: string): unknown {
   return new Database(path);
 }
 
-const SCHEMA_VERSION = 21;
+const SCHEMA_VERSION = 22;
 
 function indexExists(db: DbAdapter, name: string): boolean {
   return !!db.prepare(
@@ -949,19 +949,24 @@ function migrateSchema(db: DbAdapter): void {
     }
 
     if (currentVersion < 12) {
+      // NOTE: The original DDL used COALESCE(task_id, '') in the PRIMARY KEY
+      // expression, which is invalid SQLite syntax and causes startup errors on
+      // DBs that migrate through v12. The corrected DDL uses
+      // task_id TEXT NOT NULL DEFAULT '' with a plain column list PK. DBs that
+      // were created with the broken DDL are repaired by the v22 migration below.
       db.exec(`
         CREATE TABLE IF NOT EXISTS quality_gates (
           milestone_id TEXT NOT NULL,
           slice_id TEXT NOT NULL,
           gate_id TEXT NOT NULL,
           scope TEXT NOT NULL DEFAULT 'slice',
-          task_id TEXT DEFAULT NULL,
+          task_id TEXT NOT NULL DEFAULT '',
           status TEXT NOT NULL DEFAULT 'pending',
           verdict TEXT NOT NULL DEFAULT '',
           rationale TEXT NOT NULL DEFAULT '',
           findings TEXT NOT NULL DEFAULT '',
           evaluated_at TEXT DEFAULT NULL,
-          PRIMARY KEY (milestone_id, slice_id, gate_id, COALESCE(task_id, '')),
+          PRIMARY KEY (milestone_id, slice_id, gate_id, task_id),
           FOREIGN KEY (milestone_id, slice_id) REFERENCES slices(milestone_id, id)
         )
       `);
@@ -1185,6 +1190,52 @@ function migrateSchema(db: DbAdapter): void {
       ensureColumn(db, "memories", "structured_fields", "ALTER TABLE memories ADD COLUMN structured_fields TEXT DEFAULT NULL");
       db.prepare("INSERT INTO schema_version (version, applied_at) VALUES (:version, :applied_at)").run({
         ":version": 21,
+        ":applied_at": new Date().toISOString(),
+      });
+    }
+
+    if (currentVersion < 22) {
+      // v22: Repair quality_gates tables that were created by the broken v12
+      // migration (which used COALESCE(task_id, '') as a PK expression — invalid
+      // SQLite DDL). Those DBs have task_id nullable (dflt_value NULL, notnull 0).
+      // Rebuild the table with the correct schema, migrating existing rows via
+      // COALESCE so no data is lost.
+      const qgInfo = db.prepare("PRAGMA table_info(quality_gates)").all() as Array<Record<string, unknown>>;
+      const taskIdCol = qgInfo.find((r) => r["name"] === "task_id");
+      const needsRepair = taskIdCol && (taskIdCol["notnull"] === 0 || taskIdCol["notnull"] === "0");
+      if (needsRepair) {
+        db.exec(`
+          CREATE TABLE quality_gates_new (
+            milestone_id TEXT NOT NULL,
+            slice_id TEXT NOT NULL,
+            gate_id TEXT NOT NULL,
+            scope TEXT NOT NULL DEFAULT 'slice',
+            task_id TEXT NOT NULL DEFAULT '',
+            status TEXT NOT NULL DEFAULT 'pending',
+            verdict TEXT NOT NULL DEFAULT '',
+            rationale TEXT NOT NULL DEFAULT '',
+            findings TEXT NOT NULL DEFAULT '',
+            evaluated_at TEXT DEFAULT NULL,
+            PRIMARY KEY (milestone_id, slice_id, gate_id, task_id),
+            FOREIGN KEY (milestone_id, slice_id) REFERENCES slices(milestone_id, id)
+          )
+        `);
+        db.exec(`
+          INSERT OR IGNORE INTO quality_gates_new
+            (milestone_id, slice_id, gate_id, scope, task_id, status, verdict, rationale, findings, evaluated_at)
+          SELECT milestone_id, slice_id, gate_id, scope, COALESCE(task_id, ''), status, verdict, rationale, findings, evaluated_at
+          FROM quality_gates
+        `);
+        db.exec("DROP TABLE quality_gates");
+        db.exec("ALTER TABLE quality_gates_new RENAME TO quality_gates");
+        db.exec("CREATE INDEX IF NOT EXISTS idx_quality_gates_pending ON quality_gates(milestone_id, slice_id, status)");
+      }
+      // Ensure scope column exists on quality_gates and assessments (guard
+      // against DBs that somehow lack it after a partial migration).
+      ensureColumn(db, "quality_gates", "scope", "ALTER TABLE quality_gates ADD COLUMN scope TEXT NOT NULL DEFAULT 'slice'");
+      ensureColumn(db, "assessments", "scope", "ALTER TABLE assessments ADD COLUMN scope TEXT NOT NULL DEFAULT ''");
+      db.prepare("INSERT INTO schema_version (version, applied_at) VALUES (:version, :applied_at)").run({
+        ":version": 22,
         ":applied_at": new Date().toISOString(),
       });
     }

--- a/src/resources/extensions/gsd/journal.ts
+++ b/src/resources/extensions/gsd/journal.ts
@@ -39,7 +39,8 @@ export type JournalEventType =
   | "worktree-create-failed"
   | "worktree-skip"
   | "worktree-merge-start"
-  | "worktree-merge-failed";
+  | "worktree-merge-failed"
+  | "artifact-verification-retry";
 
 /** A single structured event in the journal. */
 export interface JournalEntry {

--- a/src/resources/extensions/gsd/tests/complete-slice.test.ts
+++ b/src/resources/extensions/gsd/tests/complete-slice.test.ts
@@ -125,9 +125,9 @@ console.log('\n=== complete-slice: schema v6 migration ===');
 
   const adapter = _getAdapter()!;
 
-  // Verify schema version is current (v21 — ADR-013 structured_fields column)
+  // Verify schema version is current (v22 — quality_gates DDL fix)
   const versionRow = adapter.prepare('SELECT MAX(version) as v FROM schema_version').get();
-  assertEq(versionRow?.['v'], 21, 'schema version should be 21');
+  assertEq(versionRow?.['v'], 22, 'schema version should be 22');
 
   // Verify slices table has full_summary_md and full_uat_md columns
   const cols = adapter.prepare("PRAGMA table_info(slices)").all();

--- a/src/resources/extensions/gsd/tests/complete-task.test.ts
+++ b/src/resources/extensions/gsd/tests/complete-task.test.ts
@@ -109,9 +109,9 @@ console.log('\n=== complete-task: schema v5 migration ===');
 
   const adapter = _getAdapter()!;
 
-  // Verify schema version is current (v21 — ADR-013 structured_fields column)
+  // Verify schema version is current (v22 — quality_gates DDL fix)
   const versionRow = adapter.prepare('SELECT MAX(version) as v FROM schema_version').get();
-  assertEq(versionRow?.['v'], 21, 'schema version should be 21');
+  assertEq(versionRow?.['v'], 22, 'schema version should be 22');
 
   // Verify all 4 new tables exist
   const tables = adapter.prepare(

--- a/src/resources/extensions/gsd/tests/ensure-db-open.test.ts
+++ b/src/resources/extensions/gsd/tests/ensure-db-open.test.ts
@@ -389,7 +389,7 @@ describe('ensure-db-open', () => {
       assert.ok(db, 'adapter should be available after ensureDbOpen');
       assert.equal(
         db.prepare('SELECT MAX(version) as version FROM schema_version').get()?.version,
-        21,
+        22,
         'legacy DB should migrate to current schema version',
       );
 

--- a/src/resources/extensions/gsd/tests/escalation.test.ts
+++ b/src/resources/extensions/gsd/tests/escalation.test.ts
@@ -348,7 +348,7 @@ test("ADR-011 P2: schema v20 fresh DB has all escalation columns on tasks + sour
   assert.ok(decCols.includes("source"), "decisions table must have source column");
 
   const version = adapter.prepare("SELECT MAX(version) as v FROM schema_version").get();
-  assert.equal(version?.["v"], 21);
+  assert.equal(version?.["v"], 22);
 });
 
 test("ADR-011 P2: findUnappliedEscalationOverride returns null when escalation_pending=1 (still pending)", (t) => {

--- a/src/resources/extensions/gsd/tests/gsd-db.test.ts
+++ b/src/resources/extensions/gsd/tests/gsd-db.test.ts
@@ -99,7 +99,7 @@ describe('gsd-db', () => {
     // Check schema_version table
     const adapter = _getAdapter()!;
     const version = adapter.prepare('SELECT MAX(version) as version FROM schema_version').get();
-    assert.deepStrictEqual(version?.['version'], 21, 'schema version should be 21');
+    assert.deepStrictEqual(version?.['version'], 22, 'schema version should be 22');
 
     // Check tables exist by querying them
     const dRows = adapter.prepare('SELECT count(*) as cnt FROM decisions').get();

--- a/src/resources/extensions/gsd/tests/issue-4540-regressions.test.ts
+++ b/src/resources/extensions/gsd/tests/issue-4540-regressions.test.ts
@@ -1,0 +1,288 @@
+/**
+ * Regression tests for issue #4540:
+ *   Bug 1 — Invalid quality_gates migration bricks gsd.db
+ *   Bug 2 — Artifact retries emit no journal event, look like stuck loops
+ */
+import { describe, test, beforeEach, afterEach } from "node:test";
+import assert from "node:assert/strict";
+import { mkdtempSync, rmSync, mkdirSync } from "node:fs";
+import { join } from "node:path";
+import { tmpdir } from "node:os";
+import { createRequire } from "node:module";
+
+import {
+  openDatabase,
+  closeDatabase,
+  insertGateRow,
+  getPendingGates,
+  _getAdapter,
+} from "../gsd-db.ts";
+
+import { emitJournalEvent, queryJournal } from "../journal.ts";
+
+const _require = createRequire(import.meta.url);
+
+// ─── helpers ─────────────────────────────────────────────────────────────────
+
+function tmpDb(): { dir: string; dbPath: string } {
+  const dir = mkdtempSync(join(tmpdir(), "gsd-4540-"));
+  return { dir, dbPath: join(dir, "gsd.db") };
+}
+
+function cleanup(dir: string): void {
+  try { rmSync(dir, { recursive: true, force: true }); } catch { /* ignore */ }
+}
+
+/**
+ * Builds a v12 database with the broken quality_gates DDL:
+ * task_id is nullable and there is no proper multi-column PK.
+ * This simulates a DB that was created before the v12 fix was applied.
+ */
+function createBrokenV12Db(dbPath: string): void {
+  const sqlite = _require("node:sqlite");
+  const db = new sqlite.DatabaseSync(dbPath);
+  db.exec("PRAGMA journal_mode=WAL");
+
+  db.exec(`CREATE TABLE schema_version (version INTEGER NOT NULL, applied_at TEXT NOT NULL)`);
+  db.prepare("INSERT INTO schema_version (version, applied_at) VALUES (?, ?)").run(12, "2025-01-01T00:00:00.000Z");
+
+  db.exec(`
+    CREATE TABLE decisions (
+      seq INTEGER PRIMARY KEY AUTOINCREMENT, id TEXT NOT NULL UNIQUE,
+      when_context TEXT NOT NULL DEFAULT '', scope TEXT NOT NULL DEFAULT '',
+      decision TEXT NOT NULL DEFAULT '', choice TEXT NOT NULL DEFAULT '',
+      rationale TEXT NOT NULL DEFAULT '', revisable TEXT NOT NULL DEFAULT '',
+      made_by TEXT NOT NULL DEFAULT 'agent', superseded_by TEXT DEFAULT NULL
+    );
+    CREATE VIEW active_decisions AS SELECT * FROM decisions WHERE superseded_by IS NULL;
+    CREATE TABLE requirements (
+      id TEXT PRIMARY KEY, class TEXT NOT NULL DEFAULT '', status TEXT NOT NULL DEFAULT '',
+      description TEXT NOT NULL DEFAULT '', why TEXT NOT NULL DEFAULT '',
+      source TEXT NOT NULL DEFAULT '', primary_owner TEXT NOT NULL DEFAULT '',
+      supporting_slices TEXT NOT NULL DEFAULT '', validation TEXT NOT NULL DEFAULT '',
+      notes TEXT NOT NULL DEFAULT '', full_content TEXT NOT NULL DEFAULT '',
+      superseded_by TEXT DEFAULT NULL
+    );
+    CREATE TABLE artifacts (
+      path TEXT PRIMARY KEY, artifact_type TEXT NOT NULL DEFAULT '',
+      milestone_id TEXT DEFAULT NULL, slice_id TEXT DEFAULT NULL, task_id TEXT DEFAULT NULL,
+      full_content TEXT NOT NULL DEFAULT '', imported_at TEXT NOT NULL DEFAULT ''
+    );
+    CREATE TABLE memories (
+      seq INTEGER PRIMARY KEY AUTOINCREMENT, id TEXT NOT NULL UNIQUE,
+      category TEXT NOT NULL, content TEXT NOT NULL, confidence REAL NOT NULL DEFAULT 0.8,
+      source_unit_type TEXT, source_unit_id TEXT, created_at TEXT NOT NULL,
+      updated_at TEXT NOT NULL, superseded_by TEXT DEFAULT NULL,
+      hit_count INTEGER NOT NULL DEFAULT 0,
+      scope TEXT NOT NULL DEFAULT 'project', tags TEXT NOT NULL DEFAULT '[]',
+      structured_fields TEXT DEFAULT NULL
+    );
+    CREATE TABLE memory_sources (
+      id TEXT PRIMARY KEY, kind TEXT NOT NULL DEFAULT '', path TEXT DEFAULT NULL,
+      imported_at TEXT NOT NULL DEFAULT '',
+      scope TEXT NOT NULL DEFAULT 'project', tags TEXT NOT NULL DEFAULT '[]'
+    );
+    CREATE TABLE memory_relations (
+      from_id TEXT NOT NULL, to_id TEXT NOT NULL, relation TEXT NOT NULL DEFAULT '',
+      PRIMARY KEY (from_id, to_id)
+    );
+    CREATE TABLE memory_processed_units (
+      unit_key TEXT PRIMARY KEY, activity_file TEXT, processed_at TEXT NOT NULL
+    );
+    CREATE TABLE milestones (
+      id TEXT PRIMARY KEY, title TEXT NOT NULL DEFAULT '', status TEXT NOT NULL DEFAULT 'active',
+      depends_on TEXT NOT NULL DEFAULT '[]', created_at TEXT NOT NULL DEFAULT '',
+      completed_at TEXT DEFAULT NULL, vision TEXT NOT NULL DEFAULT '',
+      success_criteria TEXT NOT NULL DEFAULT '[]', key_risks TEXT NOT NULL DEFAULT '[]',
+      proof_strategy TEXT NOT NULL DEFAULT '[]', verification_contract TEXT NOT NULL DEFAULT '',
+      verification_integration TEXT NOT NULL DEFAULT '',
+      verification_operational TEXT NOT NULL DEFAULT '', verification_uat TEXT NOT NULL DEFAULT '',
+      definition_of_done TEXT NOT NULL DEFAULT '[]', requirement_coverage TEXT NOT NULL DEFAULT '',
+      boundary_map_markdown TEXT NOT NULL DEFAULT ''
+    );
+    CREATE TABLE slices (
+      milestone_id TEXT NOT NULL, id TEXT NOT NULL, title TEXT NOT NULL DEFAULT '',
+      status TEXT NOT NULL DEFAULT 'pending', risk TEXT NOT NULL DEFAULT 'medium',
+      depends TEXT NOT NULL DEFAULT '[]', demo TEXT NOT NULL DEFAULT '',
+      created_at TEXT NOT NULL DEFAULT '', completed_at TEXT DEFAULT NULL,
+      full_summary_md TEXT NOT NULL DEFAULT '', full_uat_md TEXT NOT NULL DEFAULT '',
+      goal TEXT NOT NULL DEFAULT '', success_criteria TEXT NOT NULL DEFAULT '',
+      proof_level TEXT NOT NULL DEFAULT '', integration_closure TEXT NOT NULL DEFAULT '',
+      observability_impact TEXT NOT NULL DEFAULT '', sequence INTEGER DEFAULT 0,
+      replan_triggered_at TEXT DEFAULT NULL,
+      is_sketch INTEGER NOT NULL DEFAULT 0, sketch_scope TEXT NOT NULL DEFAULT '',
+      PRIMARY KEY (milestone_id, id), FOREIGN KEY (milestone_id) REFERENCES milestones(id)
+    );
+    CREATE TABLE tasks (
+      milestone_id TEXT NOT NULL, slice_id TEXT NOT NULL, id TEXT NOT NULL,
+      title TEXT NOT NULL DEFAULT '', status TEXT NOT NULL DEFAULT 'pending',
+      one_liner TEXT NOT NULL DEFAULT '', narrative TEXT NOT NULL DEFAULT '',
+      verification_result TEXT NOT NULL DEFAULT '',
+      escalation_pending INTEGER NOT NULL DEFAULT 0,
+      PRIMARY KEY (milestone_id, slice_id, id),
+      FOREIGN KEY (milestone_id, slice_id) REFERENCES slices(milestone_id, id)
+    );
+    CREATE TABLE assessments (
+      path TEXT PRIMARY KEY, milestone_id TEXT NOT NULL DEFAULT '',
+      slice_id TEXT DEFAULT NULL, task_id TEXT DEFAULT NULL,
+      status TEXT NOT NULL DEFAULT '', scope TEXT NOT NULL DEFAULT '',
+      full_content TEXT NOT NULL DEFAULT '', created_at TEXT NOT NULL DEFAULT '',
+      FOREIGN KEY (milestone_id) REFERENCES milestones(id)
+    );
+    CREATE TABLE replan_history (
+      id INTEGER PRIMARY KEY AUTOINCREMENT, milestone_id TEXT NOT NULL,
+      slice_id TEXT DEFAULT NULL, task_id TEXT DEFAULT NULL,
+      reason TEXT NOT NULL DEFAULT '', created_at TEXT NOT NULL DEFAULT ''
+    );
+    CREATE TABLE verification_evidence (
+      id INTEGER PRIMARY KEY AUTOINCREMENT, milestone_id TEXT NOT NULL DEFAULT '',
+      slice_id TEXT NOT NULL DEFAULT '', task_id TEXT NOT NULL DEFAULT '',
+      unit_type TEXT NOT NULL DEFAULT '', unit_id TEXT NOT NULL DEFAULT '',
+      evidence_type TEXT NOT NULL DEFAULT '', content TEXT NOT NULL DEFAULT '',
+      created_at TEXT NOT NULL DEFAULT '',
+      command TEXT NOT NULL DEFAULT '', verdict TEXT NOT NULL DEFAULT ''
+    );
+  `);
+
+  // Broken quality_gates: task_id nullable, no multi-column PK
+  db.exec(`
+    CREATE TABLE quality_gates (
+      milestone_id TEXT NOT NULL, slice_id TEXT NOT NULL, gate_id TEXT NOT NULL,
+      scope TEXT NOT NULL DEFAULT 'slice', task_id TEXT DEFAULT NULL,
+      status TEXT NOT NULL DEFAULT 'pending', verdict TEXT NOT NULL DEFAULT '',
+      rationale TEXT NOT NULL DEFAULT '', findings TEXT NOT NULL DEFAULT '',
+      evaluated_at TEXT DEFAULT NULL,
+      FOREIGN KEY (milestone_id, slice_id) REFERENCES slices(milestone_id, id)
+    )
+  `);
+
+  // Parent rows + gate row with NULL task_id
+  db.prepare("INSERT INTO milestones (id, title, status) VALUES (?, ?, ?)").run("M001", "Milestone 1", "active");
+  db.prepare("INSERT INTO slices (milestone_id, id, title, status, risk, depends) VALUES (?, ?, ?, ?, ?, ?)").run("M001", "S01", "Slice 1", "pending", "medium", "[]");
+  db.prepare("INSERT INTO quality_gates (milestone_id, slice_id, gate_id, scope, task_id, status) VALUES (?, ?, ?, ?, ?, ?)").run("M001", "S01", "Q3", "slice", null, "pending");
+
+  db.close();
+}
+
+// ─── Bug 1 tests ─────────────────────────────────────────────────────────────
+
+describe("Bug 1 — quality_gates migration repair (#4540)", () => {
+  let dir: string;
+  let dbPath: string;
+
+  beforeEach(() => {
+    ({ dir, dbPath } = tmpDb());
+  });
+
+  afterEach(() => {
+    closeDatabase();
+    cleanup(dir);
+  });
+
+  test("fresh DB: quality_gates task_id is NOT NULL with empty-string default", () => {
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    const cols = adapter.prepare("PRAGMA table_info(quality_gates)").all() as Array<Record<string, unknown>>;
+    const col = cols.find((c) => c["name"] === "task_id");
+    assert.ok(col, "task_id column must exist");
+    assert.equal(col["notnull"], 1, "task_id must be NOT NULL");
+    assert.equal(col["dflt_value"], "''", "task_id default must be ''");
+  });
+
+  test("fresh DB: insertGateRow with no taskId stores '' and is idempotent", () => {
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    adapter.prepare("INSERT OR IGNORE INTO milestones (id, title, status) VALUES (?, ?, ?)").run("M001", "Test", "active");
+    adapter.prepare("INSERT OR IGNORE INTO slices (milestone_id, id, title, status, risk, depends) VALUES (?, ?, ?, ?, ?, ?)").run("M001", "S01", "Slice", "pending", "medium", "[]");
+
+    insertGateRow({ milestoneId: "M001", sliceId: "S01", gateId: "Q3", scope: "slice" });
+    insertGateRow({ milestoneId: "M001", sliceId: "S01", gateId: "Q3", scope: "slice" });
+    const rows = getPendingGates("M001", "S01");
+    assert.equal(rows.length, 1);
+    assert.equal(rows[0].task_id, "");
+  });
+
+  test("v22 repair: broken v12 DB opens without error", () => {
+    createBrokenV12Db(dbPath);
+    assert.doesNotThrow(() => openDatabase(dbPath));
+  });
+
+  test("v22 repair: task_id becomes NOT NULL after healing broken v12 DB", () => {
+    createBrokenV12Db(dbPath);
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    const cols = adapter.prepare("PRAGMA table_info(quality_gates)").all() as Array<Record<string, unknown>>;
+    const col = cols.find((c) => c["name"] === "task_id");
+    assert.ok(col, "task_id column must exist after repair");
+    assert.equal(col["notnull"], 1, "task_id must be NOT NULL after repair");
+  });
+
+  test("v22 repair: existing NULL task_id row is COALESCE'd to '' during repair", () => {
+    createBrokenV12Db(dbPath);
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    const row = adapter.prepare(
+      "SELECT task_id FROM quality_gates WHERE milestone_id = 'M001' AND slice_id = 'S01' AND gate_id = 'Q3'"
+    ).get() as Record<string, unknown> | undefined;
+    assert.ok(row, "original gate row must survive the repair migration");
+    assert.equal(row["task_id"], "", "NULL task_id must be repaired to ''");
+  });
+
+  test("v22 repair: scope column present on quality_gates after open", () => {
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    const cols = adapter.prepare("PRAGMA table_info(quality_gates)").all() as Array<Record<string, unknown>>;
+    assert.ok(cols.some((c) => c["name"] === "scope"), "quality_gates.scope must exist");
+  });
+
+  test("v22 repair: scope column present on assessments after open", () => {
+    openDatabase(dbPath);
+    const adapter = _getAdapter()!;
+    const cols = adapter.prepare("PRAGMA table_info(assessments)").all() as Array<Record<string, unknown>>;
+    assert.ok(cols.some((c) => c["name"] === "scope"), "assessments.scope must exist");
+  });
+});
+
+// ─── Bug 2 tests ─────────────────────────────────────────────────────────────
+
+describe("Bug 2 — artifact-verification-retry journal event (#4540)", () => {
+  test("emitJournalEvent accepts artifact-verification-retry event type", () => {
+    const basePath = mkdtempSync(join(tmpdir(), "gsd-journal-4540-"));
+    try {
+      mkdirSync(join(basePath, ".gsd"), { recursive: true });
+      emitJournalEvent(basePath, {
+        ts: new Date().toISOString(),
+        flowId: "flow-4540",
+        seq: 1,
+        eventType: "artifact-verification-retry",
+        data: { unitType: "plan-slice", unitId: "M001/S01", attempt: 1 },
+      });
+      const entries = queryJournal(basePath, { flowId: "flow-4540" });
+      assert.equal(entries.length, 1);
+      assert.equal(entries[0].eventType, "artifact-verification-retry");
+    } finally {
+      rmSync(basePath, { recursive: true, force: true });
+    }
+  });
+
+  test("artifact-verification-retry event carries attempt count", () => {
+    const basePath = mkdtempSync(join(tmpdir(), "gsd-journal-4540b-"));
+    try {
+      mkdirSync(join(basePath, ".gsd"), { recursive: true });
+      emitJournalEvent(basePath, {
+        ts: new Date().toISOString(),
+        flowId: "flow-4540b",
+        seq: 1,
+        eventType: "artifact-verification-retry",
+        data: { unitType: "exec-slice", unitId: "M002/S02", attempt: 2 },
+      });
+      const entries = queryJournal(basePath, { flowId: "flow-4540b", eventType: "artifact-verification-retry" });
+      assert.equal(entries.length, 1);
+      const payload = entries[0].data as Record<string, unknown>;
+      assert.equal(payload["attempt"], 2);
+      assert.equal(payload["unitId"], "M002/S02");
+    } finally {
+      rmSync(basePath, { recursive: true, force: true });
+    }
+  });
+});

--- a/src/resources/extensions/gsd/tests/md-importer.test.ts
+++ b/src/resources/extensions/gsd/tests/md-importer.test.ts
@@ -363,7 +363,7 @@ test('md-importer: schema v1→v2 migration', () => {
   openDatabase(':memory:');
   const adapter = _getAdapter();
   const version = adapter?.prepare('SELECT MAX(version) as v FROM schema_version').get();
-  assert.deepStrictEqual(version?.v, 21, 'new DB should be at schema version 21');
+  assert.deepStrictEqual(version?.v, 22, 'new DB should be at schema version 22');
 
   // Artifacts table should exist
   const tableCheck = adapter?.prepare("SELECT count(*) as c FROM sqlite_master WHERE type='table' AND name='artifacts'").get();

--- a/src/resources/extensions/gsd/tests/memory-store.test.ts
+++ b/src/resources/extensions/gsd/tests/memory-store.test.ts
@@ -323,9 +323,9 @@ test('memory-store: schema includes memories table', () => {
   const viewCount = adapter.prepare('SELECT count(*) as cnt FROM active_memories').get();
   assert.deepStrictEqual(viewCount?.['cnt'], 0, 'active_memories view should exist');
 
-  // Verify schema version is 21 (ADR-013 structured_fields column included)
+  // Verify schema version is 22 (v22 quality_gates DDL fix included)
   const version = adapter.prepare('SELECT MAX(version) as v FROM schema_version').get();
-  assert.deepStrictEqual(version?.["v"], 21, 'schema version should be 21');
+  assert.deepStrictEqual(version?.["v"], 22, 'schema version should be 22');
 
   closeDatabase();
 });

